scipy/signal/_spectral_py.py

```diff
@@ -1976,29 +1999,54 @@ def _fft_helper(x, win, detrend_func, nperseg, noverlap, nfft, sides):
 
     .. versionadded:: 0.16.0
     """
+    xp = array_namespace(x)
+    x = xp.asarray(x)
     # Created strided array of data segments
     if nperseg == 1 and noverlap == 0:
-        result = x[..., np.newaxis]
+        result = x[..., xp.newaxis]
     else:
         # https://stackoverflow.com/a/5568169
         step = nperseg - noverlap
         shape = x.shape[:-1]+((x.shape[-1]-noverlap)//step, nperseg)
-        strides = x.strides[:-1]+(step*x.strides[-1], x.strides[-1])
-        result = np.lib.stride_tricks.as_strided(x, shape=shape,
-                                                 strides=strides)
+        result = xp.empty(shape, dtype=x.dtype, device=xp.device(x))
+        if "torch" in xp.__name__:
+            strides = x.stride()[:-1]+(step*x.stride()[-1], x.stride()[-1])
+            result = xp.as_strided(x, size=shape, stride=strides)
+        elif "cupy" in xp.__name__:
+            strides = x.strides[:-1]+(step*x.strides[-1], x.strides[-1])
+            result = xp.lib.stride_tricks.as_strided(x, shape=shape,
+                                                     strides=strides)
+        else:
+            # see discussion in gh-18286
+            # substantial performance penalities possible
+            # here in absence of array API support
+            for ii in range(shape[0]):
+                if len(shape) == 2:
+                    result[ii, :] = x[ii * step: (ii * step + nperseg)]
+                elif len(shape) == 3:
+                    for jj in range(shape[1]):
+                        result[ii, jj, :] = x[ii, jj * step: (jj * step + nperseg)]
+                elif len(shape) == 4:
+                    for jj in range(shape[1]):
+                        for kk in range(shape[2]):
+                            result[ii, jj, kk, :] = x[ii, jj, kk * step: (kk * step + nperseg)]
 
     # Detrend each data segment individually
     result = detrend_func(result)
 
     # Apply window by multiplication
+    # NOTE: torch device shim -- needs
```

Comment:
This should be fixed after https://github.com/pytorch/pytorch/issues/106773 I think.